### PR TITLE
In-Memory Engine: Load from ranges from region labels

### LINKED ISSUE #16655: In-Memory Engine: Load regions based on region label
## Load regions based on region-label

* Parent issue: #16417 #16141 

Load regions with region-label "cache": set to "always"

### PR DESCRIPTION
**Pull request by afeinberg**
### What is changed and how it works?

Issue Number: Close #16655 Ref #16323 Ref #16417

What's Changed:

```commit-message
If region cache memory is enabled, load based on label rules in PD.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
Tested manually with TPCC and doing:
```sql
alter table customer attributes "cache=always"; alter table district attributes "cache=always"; alter table history attributes "cache=always"; alter table item attributes "cache=always"; alter table stock attributes "cache=always"; alter table orders attributes "cache=always"; alter table order_line attributes "cache=always"; alter table new_order attributes "cache=always"; alter table warehouse attributes "cache=always";
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None

```

### CODE REVIEW

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/background.rs:161]: nit: I should probably move this to KeyRange itself.

File: components/server/src/common.rs
Comment (tonyxuqqi) [inline components/server/src/common.rs:703]: I'm thinking we can have two APIs, for is current build.
The other is build_hybrid, which has the extra parameter. 
And this parameter can be the abstraction of CacheHint
pub trait CacheHint {
    fn get_cache_hints()
    fn on_cache_hint_added()
    fn on_cache_hint_deleted()
}
Then we use pd client to implement it. The initialize of cache hint implementation can be in server.rs.

Comment (afeinberg) [reply components/server/src/common.rs:703]: Done. I did this in even a simpler way, using RangeHintService as a pure marker trait. I may later re-factor it and add members to the trait to simplify it.

**Review by tonyxuqqi — APPROVED** [review]
LGTM

**Comment by afeinberg** [comment]
Should pass clippy now. Will now add test and either fix the pd integration or add a todo for it.

File: components/region_cache_memory_engine/src/background.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/background.rs:149]: It seems to me multiple LoadRange task could run concurrently and may lead to duplicated region load?

**Review by tonyxuqqi — APPROVED** [review]

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [reply components/region_cache_memory_engine/src/background.rs:149]: Spoke offline about this. Will make this change.

**Comment by afeinberg** [comment]
Waiting for review from @SpadeA-Tang

File: components/region_cache_memory_engine/src/background.rs
Comment (SpadeA-Tang) [inline components/region_cache_memory_engine/src/background.rs:92]: Please add some comments for this as well as `PdRangeHintService`.

Comment (afeinberg) [reply components/region_cache_memory_engine/src/background.rs:101]: Done.

File: components/region_cache_memory_engine/src/engine.rs
Comment (SpadeA-Tang) [inline components/region_cache_memory_engine/src/engine.rs:439]: prepare_for_apply should be called when the relevant region begins to apply raft entries. We don't know whether there's some onging writes in the range, so if we schedule the load here (which acquires snapshot right now), ongoing writes may be lost.
For example and range k1 - k10:
In apply thread:        
write k1 in write batch and now the range is not cached nor pending, so k1 will not be buffered in the write batch of memory engine.
In BG worker:
load k1-k10, and schedule load task(acquire snapshot without write k1) and done
In apply thread:  
consume write batch

now, k1 is lost.

Comment (afeinberg) [reply components/region_cache_memory_engine/src/engine.rs:439]: @tonyxuqqi - I believe we've discussed this issue in our sync on Friday (I've brought it up.) What do you think?

Comment (afeinberg) [reply components/region_cache_memory_engine/src/engine.rs:439]: @SpadeA-Tang Yes, I also fear this scenario. Any objection if I create an issue to handle it in a follow on PR, along with a test that will replicate it? (Don't want this to grow too large.)

Comment (afeinberg) [reply components/region_cache_memory_engine/src/engine.rs:439]: > write k1 in write batch and now the range is not cached nor pending, so k1 will not be buffered in the write batch of memory engine.

Writing k1 needs to happen to rocksdb, which generates a new sequence number. Wouldn't we thus not be able to read a version containing k1 as we don't have a snapshot with that sequence number?

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:439]: I think we can handle the issue of not receiving new raft command in the following PR with deleting this line of code.

Comment (afeinberg) [reply components/region_cache_memory_engine/src/engine.rs:439]: 'so i see now: earlier I was buffering all writes in the batch and then making the decision about whether to write them in the callback. this has changed. 

re: moving this to region's apply. 

how would prepare_for_apply know what the relevant range to load is if the cached range is greater than one specific region?  this is the current behavior if we do "ALTER TABLE set attributes" for example. this would require an additional way to communicate that specific range: keep in mind, region_labels can work with labels for an entire table and not just one specific region. 

note that this PR is still fairly limited: in this case, the loading only happens instantly and only when a region is labeled manually via PD. i think it's safe to do it here for this PR.

Comment (afeinberg) [reply components/region_cache_memory_engine/src/engine.rs]: > I think we can handle the issue of not receiving new raft command in the following PR with deleting this line of code.

Done.

**Review by SpadeA-Tang — APPROVED** [review]

**Pull request merged**
